File: Graph/java/IslandCount.java
import java.util.HashSet;
import java.util.Set;

public class IslandCount {
  public static int count(char[][] grid) {
    Set<String> visited = new HashSet<>();
    int total = 0;

    for (int r = 0; r < grid.length; r++) {
      for (int c = 0; c < grid[0].length; c++) {
        if (explore(grid, r, c, visited)) {
          total++;
        }
      }
    }

    return total;
  }

  private static boolean explore(char[][] grid, int row, int col, Set<String> visited) {
    if (!GraphUtils.inBounds(grid, row, col) || grid[row][col] == 'W') {
      return false;
    }

    String key = row + "," + col;
    if (!visited.add(key)) {
      return false;
    }

    for (int[] delta : GraphUtils.DIRECTIONS) {
      explore(grid, row + delta[0], col + delta[1], visited);
    }

    return true;
  }
}

File: Graph/ts/islandCount.ts
import { Grid, inBounds, posKey } from "./grid";

export function islandCount(grid: Grid): number {
  const visited = new Set<string>();
  let count = 0;

  for (let r = 0; r < grid.length; r += 1) {
    for (let c = 0; c < grid[0].length; c += 1) {
      if (explore(grid, r, c, visited)) {
        count += 1;
      }
    }
  }

  return count;
}

function explore(grid: Grid, r: number, c: number, visited: Set<string>): boolean {
  if (!inBounds(grid, r, c)) return false;
  if (grid[r][c] === "W") return false;

  const key = posKey(r, c);
  if (visited.has(key)) return false;
  visited.add(key);

  explore(grid, r - 1, c, visited);
  explore(grid, r + 1, c, visited);
  explore(grid, r, c - 1, visited);
  explore(grid, r, c + 1, visited);

  return true;
}
